# Becoming a successful Travel Agent

In this assignment, we will devise a cheapest flight booking algorithm that beats the some of the worlds best flight booking sites. Specifically, we will find cheapest flights to NYC from various airports with layovers as time steps. A simgle time step optimization means direct flight. Two step means one layover in between. Three steps means 2 layovers in between and so on.

Here our states will be airports and control will be choosing a specific airline to a specific airport. Therefore our state update function returns destination of the control chosen. The cost function returns cost of that flight. Note that cost function is additive. For example, if we decide to go BOM-LHR-NYC, the total cost = cost of BOM-LHR + cost of LHR-NYC. To see if our algorithm works, for cost function, we will use actual flight prices for 13th January 2018 from KAYAK website.

In [114]:
airports = ['BOM', 'NYC', 'DXB', 'LHR', 'FRA', 'DOH']
airlines = ['AIR_INDIA', 'BRITISH_AIRWAYS', 'EMIRATES', 
            'QATAR_AIRWAYS', 'LUFTHANSA']

bom = {'AIR_INDIA': [0, 1300, 198, 641, 149000000, 2326],
       'BRITISH_AIRWAYS': [0, 149000000, 149000000, 598, 
                           149000000, 149000000],
       'EMIRATES': [0, 149000000, 269, 149000000, 149000000, 149000000],
       'QATAR_AIRWAYS': [0, 149000000, 149000000, 149000000, 149000000, 925],
       'LUFTHANSA': [0, 149000000, 149000000, 149000000, 1371, 149000000]}

nyc = {'AIR_INDIA': [849, 0, 149000000, 390, 149000000, 149000000],
       'BRITISH_AIRWAYS': [149000000, 0, 149000000, 1664, 
                           149000000, 149000000],
       'EMIRATES': [149000000, 0, 861, 149000000, 149000000, 149000000],
       'QATAR_AIRWAYS': [149000000, 0, 149000000, 149000000, 149000000, 1176],
       'LUFTHANSA': [149000000, 0, 149000000, 1667, 2877, 149000000]}

dxb = {'AIR_INDIA': [112, 149000000, 0, 149000000, 149000000, 149000000],
       'BRITISH_AIRWAYS': [149000000, 149000000, 0, 725, 149000000, 149000000],
       'EMIRATES': [128, 1128, 0, 823, 772, 149000000],
       'QATAR_AIRWAYS': [149000000, 149000000, 0, 149000000, 
                         149000000, 149000000],
       'LUFTHANSA': [149000000, 149000000, 0, 149000000, 586, 149000000]}

lhr = {'AIR_INDIA': [405, 392, 149000000, 0, 149000000, 149000000],
       'BRITISH_AIRWAYS': [965, 149000000, 596, 0, 198, 819],
       'EMIRATES': [149000000, 149000000, 617, 0, 149000000, 149000000],
       'QATAR_AIRWAYS': [149000000, 149000000, 149000000, 0, 149000000, 950],
       'LUFTHANSA': [149000000, 1764, 149000000, 0, 281, 149000000]}

fra = {'AIR_INDIA': [975, 149000000, 149000000, 149000000, 0, 149000000],
       'BRITISH_AIRWAYS': [149000000, 149000000, 149000000, 206, 0, 149000000],
       'EMIRATES': [149000000, 149000000, 590, 149000000, 0, 149000000],
       'QATAR_AIRWAYS': [149000000, 149000000, 149000000, 149000000, 0, 558],
       'LUFTHANSA': [2662, 723, 149000000, 295, 0, 149000000]}

doh = {'AIR_INDIA': [166, 149000000, 149000000, 149000000, 149000000, 0],
       'BRITISH_AIRWAYS': [149000000, 149000000, 149000000, 715, 149000000, 0],
       'EMIRATES': [149000000, 149000000, 149000000, 149000000, 149000000, 0],
       'QATAR_AIRWAYS': [223, 1222, 149000000, 771, 616, 0],
       'LUFTHANSA': [149000000, 149000000, 149000000, 149000000, 149000000, 0]}

price_matrix = {'BOM': bom, 'NYC': nyc, 'DXB': dxb, 
                'LHR': lhr, 'FRA': fra, 'DOH': doh}


# First lets assume a 1 step process i.e. no layover
# The minimum cost to reach NYC from each possible state (current airport)
# is the minimum cost amongst all airlines going from current airport to NYC
# For example, in one step, minimum cost to reach from BOM to NYC will be
# minimum of second column of BOM matrix 
# (each row is prices of particular airline to reach different airports)

# Write a function get_direct_lowest_cost_airline(departure, arrival)
# which takes input departure and arrival airports and returns the cost of direct flight

def get_direct_lowest_cost_airline(departure, arrival):
    minimum_cost_to_reach_arrival = 149000000
    for airline, prices in price_matrix[departure].items():
        if prices[airports.index(arrival)] < minimum_cost_to_reach_arrival:
            minimum_cost_to_reach_arrival = prices[airports.index(arrival)]
    return minimum_cost_to_reach_arrival


# Write a function that returns minimum one step (direct) flight cost to NYC


# Write a function that returns 2 step (one layover) flight cost to NYC


# Write a function that returns optimal cost (less than N stop) flight to NYC from all airports

def optimal_policy(to, max_stops):
    minimum_cost_for_zero_stops = [get_direct_lowest_cost_airline(d, to) 
                                   for d in airports]
    if max_stops == 1:
        return minimum_cost_for_zero_stops
    stops = 1
    minimum_cost_for_one_less_stop = minimum_cost_for_zero_stops
    while stops <= max_stops:
        minimum_cost_airlines_for_all_paths = []
        for layover in airports:
            index_of_layover_airport = airports.index(layover)
            minimum_cost_to_reach_next_stop_from_layover = \
                minimum_cost_for_one_less_stop
            minimum_cost_to_reach_layover_airport_by_departures = []
            for departure in airports:
                temp = minimum_cost_to_reach_next_stop_from_layover[
                           index_of_layover_airport] + \
                       get_direct_lowest_cost_airline(departure, layover)
                minimum_cost_to_reach_layover_airport_by_departures.append(temp)
            minimum_cost_airlines_for_all_paths.append(
                minimum_cost_to_reach_layover_airport_by_departures)

        minimum_cost_for_departure_to_arrival = [
            min([r[i] for r in minimum_cost_airlines_for_all_paths]) for i in
                                                 range(len(airports))]
        minimum_cost_for_one_less_stop = minimum_cost_for_departure_to_arrival
        stops += 1
    return minimum_cost_for_one_less_stop


cost_to_reach_NYC = optimal_policy('NYC', 2)
for d, c in zip(airports, cost_to_reach_NYC):
    print("Cheapest flight from {} to NYC costs ${}".format(d, c))

Cheapest flight from BOM to NYC costs $990
Cheapest flight from NYC to NYC costs $0
Cheapest flight from DXB to NYC costs $1102
Cheapest flight from LHR to NYC costs $392
Cheapest flight from FRA to NYC costs $598
Cheapest flight from DOH to NYC costs $1107


Itinerary calculated by our travel agent -   
BRITISH AIRWAYS from BOM to LHR - \$ 598  
AIR INDIA from LHR to NYC - \$ 392  
Total cost - \$ 990  

To give you a glimpse of how successful our algorithm was, lets look at what KAYAK, a wellknown flight booking platform, recommends for lowest cost flights from BOM to NYC for same dates for same set of airlines

We see that our algorithm has saved alomst \$ 100 for our customers. Now go ahead and build your own cheapest flight booking website powered by dynamic programming!

![title](airline1.png)
![title](airline2.png)